### PR TITLE
feat(metadata): Add new metadata query api class

### PR DESCRIPTION
**Pull request by patrickpaul**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @patrickpaul has signed the CLA. Thanks for the pull request!

File: src/api/MetadataQuery.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        columns: List<ColumnType>,
+        conditions: List<ConditionType>,
+        orderByColumn: ColumnType,
+        orderDirection: SortOrderType,
+        connector: ConnectorType,
+    ) => {
+        const argumentList = {};
+        const conditionsList = [];
+
+        conditions.forEach(({ columnId, operator, values }, conditionIndex) => {
```
Comment (jstoffan) [inline src/api/MetadataQuery.js]: The `forEach` iterator function is doing quite a bit. Consider breaking it into multiple other functions for readability/testability.

File: src/api/MetadataQuery.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            const operatorSymbol = operator;
+            const targetColumn = columns.filter(column => column.id === columnId).get(0);
+            if (!targetColumn) {
+                return;
+            }
+            const fieldKey = targetColumn.property;
+            conditionsList.push(`${scope}.${templateKey}.${fieldKey} ${operatorSymbol} :arg_${conditionIndex}`);
+
+            if (!Number.isNaN(values[0])) {
```
[the remark this answers is not in the document]
Comment (patrickpaul) [reply src/api/MetadataQuery.js]: Should we not lean on transpiler? We use `Number.isNaN` throughout the uploads code. If we switch to lodash I'd want to change all instances.

[the remark this answers is not in the document]
Comment (jstoffan) [reply src/api/MetadataQuery.js]: I don't believe such code is transformed by Babel and would need to be polyfilled. You can check this yourself by running the `build:prod:es` script and checking the `/es` directory.

[the remark this answers is not in the document]
Comment (jstoffan) [reply src/api/MetadataQuery.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN#Browser_compatibility

File: src/api/MetadataQuery.js (hunk 1 above)
Comment (patrickpaul) [reply src/api/MetadataQuery.js]: Agreed but going to punt on this until the section is reworked (notice all the lingering TODOs).

File: src/api/MetadataQuery.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+ */
+import { List } from 'immutable';
+
+import type { ColumnType, ConditionType, ConnectorType, SortOrderType } from '../features/query-bar/flowTypes';
+import Base from './Base';
+import { ERROR_CODE_METADATA_QUERY } from '../constants';
+
+class MetadataQuery extends Base {
+    /**
+     * URL for recents api
```
Comment (macrotim) [inline src/api/MetadataQuery.js]: Is this accurate?

File: src/api/MetadataQuery.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        marker: string,
+        successCallback: Function,
+        errorCallback: ElementsErrorCallback,
+    ): Promise<void> {
+        this.errorCode = ERROR_CODE_METADATA_QUERY;
+        this.successCallback = successCallback;
+        this.errorCallback = errorCallback;
+
+        try {
+            const { data }: { data: MetadataBoxItemCollection } = await this.xhr.post({
```
Comment (jstoffan) [inline src/api/MetadataQuery.js]: Should this use `this.post` rather than accessing the inner XHR property?

File: src/api/MetadataQuery.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        scope: string,
+        templateKey: string,
+        columns: List<ColumnType>,
+        conditions: List<ConditionType>,
+        orderByColumn: ColumnType,
+        orderDirection: SortOrderType,
+        connector: ConnectorType,
+    ) => {
+        const queryParams = {}; // Hashmap of arguments used in conditions
+        const conditionsList = []; // Array of conditions joined (by connector) to form the final query predicate
```
Comment (macrotim) [inline src/api/MetadataQuery.js]: It's hard to differentiate the purpose of `conditions` vs `conditionsList`. 

Alternative names might be `conditionStrings`, `stringifiedConditions`, `conditionsRightBeforeConnectorJoin`. Open to suggestions.

File: src/api/MetadataQuery.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        conditions.forEach(({ columnId, operator: operatorSymbol, values }, conditionIndex) => {
+            const targetColumn = columns.filter(column => column.id === columnId).get(0);
+            if (!targetColumn) {
+                return;
```
Comment (macrotim) [inline src/api/MetadataQuery.js]: If it's an invalid state, I think `throw new Error("Invalid state")` is more appropriate.

File: src/api/MetadataQuery.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        ancestorFolderId: string,
+        queryPostBody: Object,
+        limit: number,
+        marker: string,
+        successCallback: Function,
+        errorCallback: ElementsErrorCallback,
+    ): void {
+        this.errorCode = ERROR_CODE_METADATA_QUERY;
+
+        this.post({
```
Comment (macrotim) [inline src/api/MetadataQuery.js]: I wonder what OAuth scope is granted to this.post(). Let's talk more in person.

File: src/api/MetadataQuery.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            id: ancestorFolderId,
```
Comment (macrotim) [inline src/api/MetadataQuery.js]: What's the purpose of id? 

If needed to distinguish between queries, `ancestoryFolderId` is not sufficient.

File: src/api/MetadataQuery.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            url: this.getUrl(),
+            data: {
+                ancestorFolderId,
+                limit,
+                marker,
+                ...queryPostBody,
+            },
+            successCallback,
```
Comment (macrotim) [inline src/api/MetadataQuery.js]: Promise is preferred over callback. Is it not available?

File: src/api/MetadataQuery.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+     * @return {string} base url for files
+     */
+    getUrl(): string {
+        return `${this.getBaseApiUrl()}/metadata_queries/execute`;
+    }
+
+    /**
+     * Utility to build query post body from conditions
+     *
+     * @param {string} scope - metadata scope, usually global or enterprise_
```
Comment (ChenCodes) [inline src/api/MetadataQuery.js]: Is this a typo at the end of enterprise_?

File: src/api/MetadataQuery.js [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+     * @param {string} templateKey - the name of the metadata template
+     * @param {List<ColumnType>} columns - the list of columns/fields in the template
+     * @param {List<ConditionType>} conditions - the list of query conditions
+     * @param {ColumnType} orderByColumn - the column to order by
+     * @param {SortOrderType} orderDirection - the typed sort order ('ASC', 'DESC')
+     * @param {ConnectorType} connector connector for query ('AND' / 'OR')
+     * @return {void}
+     */
+
+    createQueryPostBody = (
```
Comment (jeremypress) [inline src/api/MetadataQuery.js]: for maintenance sake this function seems worthy of a unit test.

**Comment by patrickpaul** [comment]
Closing this for now as it is no longer needed. Will likely revisit in the future.

**Pull request closed without merging**